File: Data Structures & Algorithms/longest-palindromic-substring/submission-0.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[0] * n for _ in range(n)]
        res = [0, 0]
        
        for i in range(n):
            dp[i][i] = True
        for i in range(n - 1):
            dp[i][i + 1] = s[i] == s[i + 1]
            if dp[i][i + 1]: res = [i, i + 1]
        
        for diff in range(2, n):
            for i in range(n - diff):
                j = i + diff
                if s[i] == s[j] and dp[i + 1][j - 1]:
                    dp[i][j] = True
                    res = [i, j]
                else:
                    dp[i][j] = False
        
        return s[res[0]: res[1] + 1]

File: Data Structures & Algorithms/palindromic-substrings/submission-0.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int l = s.length();
        vector<vector<bool>> dp(l, vector<bool>(l, false));
        int res = 0;

        for (int i = 0; i < l; i++) {
            dp[i][i] = 1;
            res++;
        }
        for (int i = 0; i < l - 1; i++) {
            dp[i][i + 1] = s[i] == s[i + 1];
            if (dp[i][i + 1]) {
                cout << i << ", " << i + 1 << endl;
                res++;
            }
        }

        for (int diff = 2; diff < l; diff++) {
            for (int i = 0; i < l - diff; i++) {
                int j = i + diff;
                cout << i << ", " << j;
                dp[i][j] = (s[i] == s[j]) && (dp[i + 1][j - 1]);
                if (dp[i][j]) {
                    res++;
                } 
            }
        }

        return res;
    }
};
